### ISSUE TITLE
B023 False positives

### ISSUE THREAD

**Issue by TomFryers** [charliermarsh/ruff#1686]
False positive B023 errors are quite common with `filter`, `reduce`, `key=`, `return` and `map`. For instance, this code
```python
unsorted_things = [(["foo", "bar"], 0), (["foobar", "barfoo"], 4), (["baz", "foo", "moo"], 1)]
for unsorted_thing, letter in unsorted_things:
    unsorted_thing.sort(key=lambda x: (x[letter], len(x)))
```
is fine, but it currently gets flagged as B023.

This was raised in Flake8 Bugbear in issue [303](https://github.com/PyCQA/flake8-bugbear/pull/303) (and [305](https://github.com/PyCQA/flake8-bugbear/pull/305)) and fixed in [22.10.27](https://github.com/PyCQA/flake8-bugbear#221027).

Ignoring these does cause some false negatives, for instance if the `filter` or `map` is not evaluated before the next iteration, but this was deemed acceptable in the Bugbear issue.

**Comment by charliermarsh** [comment]
Ah cool. We can fix this! Thanks for filing.
